routing/routing_session.cpp

```diff
@@ -304,20 +309,73 @@ RoutingSession::State RoutingSession::OnLocationPositionChanged(GpsInfo const &
       SetState(OnRoute);
 
       // Warning signals checks
-      if (m_routingSettings.m_speedCameraWarning && !m_speedWarningSignal)
+      if (m_routingSettings.m_speedCameraWarningEnabled)
       {
-        double const warningDistanceM = max(kSpeedCameraMinimalWarningMeters,
-                                            info.m_speed * kSpeedCameraWarningSeconds);
-        SpeedCameraRestriction cam(0, 0);
-        double const camDistance = GetDistanceToCurrentCamM(cam, dataSource);
-        if (kInvalidSpeedCameraDistance != camDistance && camDistance < warningDistanceM)
+        auto const passedDistance = m_route->GetCurrentDistanceFromBeginMeters();
+
+        // Step 1. Process warned cameras.
+        while (!m_warnedSpeedCameras.empty())
+        {
+          auto const & oldestCamera = m_warnedSpeedCameras.front();
+          double const distBetweenCameraAndCurrentPos = passedDistance - oldestCamera.m_distFromBegin;
+          if (distBetweenCameraAndCurrentPos > SpeedCameraOnRoute::kDangerousZoneMeters)
+            m_warnedSpeedCameras.pop();
+          else
+            break;
+        }
+
+        // We will turn off big red button only after all cameras, that were be warned, exhausted.
+        if (m_warnedSpeedCameras.empty())
+          m_showWarningAboutSpeedCam = false;
+
+        // Step 2. Check cached cameras.
+        if (!m_cachedSpeedCameras.empty())
+        {
+          auto const & closestSpeedCamera = m_cachedSpeedCameras.front();
+          if (closestSpeedCamera.m_distFromBegin < passedDistance)
+          {
+            PassCameraToWarned();
+          }
+          else
+          {
+            auto const distanceToCamera = closestSpeedCamera.m_distFromBegin - passedDistance;
+            if (closestSpeedCamera.IsDangerous(distanceToCamera, info.m_speed))
+              ProcessCameraWarning();
+          }
+        }
+
+        // Step 3. Find new cameras, to be cached.
+        size_t lastChecked = m_lastCheckedSpeedCameraIndex;
+        double distToPrevSegment =
+            m_route->GetRouteSegments()[lastChecked].GetDistFromBeginningMeters();
+        double distFromCurPosToLatestCheckedSegment = distToPrevSegment - passedDistance;
```

Comment:
const? dist in meters?

Reply:
add suffix `M`. 
const - no, we update it